text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,243 @@
+
+* The run image layers are persistented as part of the app image, and can be reused on subsequent builds.
+* The build image layers will be persisted as part of the cache, and can be reused on subsequent builds.
+* Any paths that are excluded using `launch.toml` and set to be cached will be stored in the cache, and can be reused on subsequent builds.
+
+## Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral. In effect, we are saying that the use of root buildpacks breaks rebase.
+
+When the base image has not changed
+1. If the buildpacks are configure as idempotent (the default) load the previous layers onto the base images.
+1. Run the buildpacks, creating an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+When there is an update to either the build or run base images
+1. Pull the new base image(s)
+1. Run the root buildpacks against the new image(s) without loading previous layers, and create an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+## Example: Apt Buildpack
+
+A buildpack that reads an `apt.toml` file to install an arbitrary list of packages would have a `buildpack.toml` like this:
+
+```toml
+[buildpack]
+id = "example/apt"
+privileged = true
+```
+
+An end-user's `apt.toml` might look like this:
+
+```toml
+[build]
+packages = [
+  "libpq-dev"
+]
+
+[run]
+packages = [
+  "libpq",
+  "ffmpeg"
+]
+```
+
+The `bin/build` would look like:
+
+```bash
+#!/usr/bin/env bash
+
+apt update
+
+for package in $(cat apt.toml | yj -t | jq -r ".$CNB_STACK_TYPE.packages | .[]"); do
+    apt install $package
+done
+
+cat << EOF > launch.toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
+EOF
+```
+
+# Alternatives
+
+* https://github.com/buildpacks/rfcs/pull/23
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Requires kaniko (or a similar tool) for snapshotting
+- Fully rebasing apps with package dependencies becomes many orders of magnitude slower, due to upgrade needs
+- Reliability of rebasing apps with package dependencies degrades and network and other external factors may break the process.
+- Some buildpack users will not want to give root access to arbitrary buildpacks. Instead they may want to selectively whitelist certain root buildpacks.
+
+# Questions
+[questions]: #questions
+
+* Should we allow non-root buildpacks to run in combination with root buildpacks?
+* How do the packages installed by a root buildpack related (or not related) to mixins
+    * That is, can you use a root buildpack to install a package that satisfies the requirements of a buildpacks' required mixins?
```

Comment:
Why wouldn't this be a positive outcome?

Reply:
I think it would be positive, but it's an open question because we aren't defining how it would work.